x-pack/plugins/fleet/server/routes/agent/upgrade_handler.ts

```diff
@@ -48,7 +49,25 @@ export const postAgentUpgradeHandler: RequestHandler<
       },
     });
   }
+
   const agent = await getAgentById(esClient, request.params.agentId);
+
+  let sourceUri;
+  if (requestedSourceUri) {
+    sourceUri = requestedSourceUri;
+  } else {
+    try {
+      sourceUri = await getSourceUriForAgent(soClient, agent);
```

Comment:
Do we need this? if the source URI is not overwritten in the route it's already configured in the agent policy right?

Reply:
yes if we don't overwrite it it's already in the agent policy, but I added this to write the correct `source_uri `directly in `fleet_actions`. This function gets the value configured in the agent policy and if none is found gets the global default one.

Reply:
In this case we need to do the same for bulk upgrades? or should the `source_uri` should always be configured in the policy and only override if needed in the routes?

Reply:
I added the same call to bulk upgrades as well [here](https://github.com/elastic/kibana/blob/a7132206f23a189f7e7bce022eccac90b598a2b0/x-pack/plugins/fleet/server/services/agents/upgrade.ts#L174-L189), so the behaviour is aligned. 
Writing the `source_uri` to the actions was in the requirements for this feature, but now that the elastic agent team is doing the work to reload the policy at runtime maybe it won't be needed anymore. @kpollich what do you think?

Reply:
In bulk upgrade you fetch `getDefaultSourceUri`, not `getSourceUriForAgent`, is that correct?

We should avoid querying source uri for each agent in bulk upgrade, because it would query agent policy one by one, which is going to be slow.

Reply:
yes it's correct because in bulk upgrades I'm just getting the default value, we don't want to query each agent exactly because it would be slow and would have an impact on the scalability. There was a discussion about the bulk upgrades [here](https://github.com/elastic/kibana/issues/135579).

Reply:
> but now that the elastic agent team is doing the work to reload the policy at runtime maybe it won't be needed anymore

If it's not needed we can simplify a lot that, by just relying on the policy, and only allow source URI in the handler only to override the value.

Reply:
@nchaulet @kpollich For the case when`source_uri` is  passed in `POST agent/upgrade` endpoint, I'm removing the code that writes to the actions and I'm replacing it with a call to `AgentService.Update`. This way it will be written to the agent policy.
However I'm not sure how to manage it in the case of bulk upgrades, as I should update all the agent policies for every single agent in the batch. This would be expensive and quite slow.
I'm thinking that it would be better to remove altogether `source_uri` in case of bulk upgrades. What do you think?

Reply:
> @nchaulet @kpollich For the case whensource_uri is passed in POST agent/upgrade endpoint, I'm removing the code that writes to the actions and I'm replacing it with a call to AgentService.Update. This way it will be written to the agent policy.
However I'm not sure how to manage it in the case of bulk upgrades, as I should update all the agent policies for every single agent in the batch. This would be expensive and quite slow.
I'm thinking that it would be better to remove altogether source_uri in case of bulk upgrades. What do you think?

@criamico I do not think you should update the agent policy in the case `source_uri` is passed is just a one time override in the action, for single or bulk upgrade action